staging/src/k8s.io/component-base/metrics/metric.go

```diff
@@ -221,15 +221,16 @@ var noop = &noopMetric{}
 
 type noopMetric struct{}
 
-func (noopMetric) Inc()                              {}
-func (noopMetric) Add(float64)                       {}
-func (noopMetric) Dec()                              {}
-func (noopMetric) Set(float64)                       {}
-func (noopMetric) Sub(float64)                       {}
-func (noopMetric) Observe(float64)                   {}
-func (noopMetric) ObserveWithWeight(float64, uint64) {}
-func (noopMetric) SetToCurrentTime()                 {}
-func (noopMetric) Desc() *prometheus.Desc            { return nil }
-func (noopMetric) Write(*dto.Metric) error           { return nil }
-func (noopMetric) Describe(chan<- *prometheus.Desc)  {}
-func (noopMetric) Collect(chan<- prometheus.Metric)  {}
+func (noopMetric) Inc()                                           {}
+func (noopMetric) Add(float64)                                    {}
+func (noopMetric) Dec()                                           {}
+func (noopMetric) Set(float64)                                    {}
+func (noopMetric) Sub(float64)                                    {}
+func (noopMetric) Observe(float64)                                {}
+func (noopMetric) ObserveWithExemplar(float64, prometheus.Labels) {}
```

Comment:
I'm revisiting this after a [good while](https://github.com/prometheus/prometheus/pull/12738) but IIRC from a discussion with @logicalhan around exemplar support for counters, we want to modify the [existing methods](https://github.com/kubernetes/kubernetes/pull/119949/files#diff-e318d079505983653611023d8593be867ebd92d90f9522093188b25dbeca95a3R103) implicitly (without modifying their signatures) to support exemplars OOTB.

Additionally, my understanding is that we don't want to make any component [directly dependent](https://github.com/kubernetes/kubernetes/pull/124013/files#diff-c4f07ee44a4626c817dd95cf30e147a2125ea923804be5c3608fb4a845fc1c20R31) on `prometheus/client_golang`, but incorporate that behavior into `component-base` and have them import that instead, in an effort to be able to offset any breaking changes and not having to send patches upstream (and catering to any extra use-cases specific to `k/k`).

All components themselves will be responsible for ["attaching" exemplars within their contexts](https://github.com/kubernetes/kubernetes/pull/124013/files#diff-c4f07ee44a4626c817dd95cf30e147a2125ea923804be5c3608fb4a845fc1c20R31), if they want to utilize this feature-set, but we won't have to make any explicit changes in any component for this feature-set to work. However, the SIG can help components benefit from this by patching their contexts for them, but I believe that is out of the scope of this effort (adding exemplar support to `component-base`).

Reply:
Thanks for the pointers @rexagod. The initial version of this PR was mostly a POC to verify that we are able to see exemplars after-all in metrics, hence admittedly had quite a lot of hacky stuff going on at first.

I modified this PR now, following what you are doing for counters in https://github.com/kubernetes/kubernetes/pull/119949, though it still needs to use the `component-base.metrics.exemplarMetric` that you have introduced, so I'll wait for that PR to merge first and make those changes then.

Additionally I am wondering, if we need to introduce Add()/Observe() methods for CounterVec/HistogramVec too?